lib/spack/spack/cmd/checksum.py

```diff
@@ -30,93 +32,133 @@
-    # archive after it is downloaded.
+    :param dict url_dict: A dictionary of the form: version -> URL
+    :param str name: The name of the package
+    :param callable first_stage_function: Function to run on first staging area
+    :param bool keep_stage: Don't clean up staging area when command completes
+
+    :returns: A multi-line string containing versions and corresponding hashes
+    :rtype: str
+    """
     first_stage_function = kwargs.get('first_stage_function', None)
     keep_stage = kwargs.get('keep_stage', False)
 
+    sorted_versions = sorted(url_dict.keys(), reverse=True)
+
+    # Find length of longest string in the list for padding
+    max_len = max(len(str(v)) for v in sorted_versions)
+    num_ver = len(sorted_versions)
+
+    tty.msg("Found {0} version{1} of {2}:".format(
+            num_ver, '' if num_ver == 1 else '', name),
+            "",
+            *spack.cmd.elide_list(
+                ["{0:{1}}  {2}".format(v, max_len, url_dict[v])
+                 for v in sorted_versions]))
+    print()
+
+    archives_to_fetch = tty.get_number(
+        "How many would you like to checksum?", default=1, abort='q')
+
+    if not archives_to_fetch:
+        tty.die("Aborted.")
+
+    versions = sorted_versions[:archives_to_fetch]
+    urls = [url_dict[v] for v in versions]
+
     tty.msg("Downloading...")
-    hashes = []
+    version_hashes = []
     i = 0
     for url, version in zip(urls, versions):
         try:
             with Stage(url, keep=keep_stage) as stage:
+                # Fetch the archive
                 stage.fetch()
                 if i == 0 and first_stage_function:
+                    # Only run first_stage_function the first time,
+                    # no need to run it every time
                     first_stage_function(stage, url)
 
-                hashes.append((version, spack.util.crypto.checksum(
+                # Checksum the archive and add it to the list
+                version_hashes.append((version, spack.util.crypto.checksum(
                     hashlib.md5, stage.archive_file)))
                 i += 1
-        except FailedDownloadError as e:
-            tty.msg("Failed to fetch %s" % url)
+        except FailedDownloadError:
+            tty.msg("Failed to fetch {0}".format(url))
         except Exception as e:
-            tty.msg('Something failed on %s, skipping.\n    (%s)' % (url, e))
+            tty.msg("Something failed on {0}, skipping.".format(url),
+                    "  ({0})".format(e))
 
-    return hashes
+    if not version_hashes:
+        tty.die("Could not fetch any versions for {0}".format(name))
+
+    # Find length of longest string in the list for padding
+    max_len = max(len(str(v)) for v, h in version_hashes)
+
+    # Generate the version directives to put in a package.py
+    version_lines = "\n".join([
+        "    version('{0}', {1}'{2}')".format(
+            v, ' ' * (max_len - len(str(v))), h) for v, h in version_hashes
+    ])
```

Comment:
Previously, `spack checksum tau` would result in:
```
    version('2.26', '2af91f02ad26d5bf0954146c56a8cdfa')
    version('2.25.2', 'f5e542d41eb4a7daa6241e5472f49fd7')
    version('2.25.1.1', 'f2baae27c5c024937566f33339826d7c')
    version('2.25.1', 'b9783f9bbe2862254bfdd208735241b6')
    version('2.25', '46cd48fa3f3c4ce0197017b3158a2b43')
```
while
```
$ spack create --force https://www.cs.uoregon.edu/research/tau/tau_releases/tau-2.25.tar.gz
```
would result in:
```
    version('2.26'    , '2af91f02ad26d5bf0954146c56a8cdfa')
    version('2.25.2'  , 'f5e542d41eb4a7daa6241e5472f49fd7')
    version('2.25.1.1', 'f2baae27c5c024937566f33339826d7c')
    version('2.25.1'  , 'b9783f9bbe2862254bfdd208735241b6')
    version('2.25'    , '46cd48fa3f3c4ce0197017b3158a2b43')
```
The latter isn't even PEP 8 compliant, and would raise problems during the flake8 tests. Now, both result in:
```
    version('2.26',     '2af91f02ad26d5bf0954146c56a8cdfa')
    version('2.25.2',   'f5e542d41eb4a7daa6241e5472f49fd7')
    version('2.25.1.1', 'f2baae27c5c024937566f33339826d7c')
    version('2.25.1',   'b9783f9bbe2862254bfdd208735241b6')
    version('2.25',     '46cd48fa3f3c4ce0197017b3158a2b43')
```